packages/babel-plugin-transform-template-literals/src/index.js

```diff
@@ -83,19 +83,27 @@ export default declare((api, options) => {
           callExpressionInput.push(t.arrayExpression(raws));
         }
 
-        const init = t.callExpression(helperId, callExpressionInput);
-        annotateAsPure(init);
-        init._compact = true;
-        scope.push({
-          id: templateObject,
-          init,
-          // This ensures that we don't fail if not using function expression helpers
-          _blockHoist: 1.9,
-        });
+        const callExpression = t.callExpression(helperId, callExpressionInput);
+        annotateAsPure(callExpression);
+        callExpression._compact = true;
+
+        const lazyLoad = t.logicalExpression(
```

Comment:
I wonder if we'll hit function size deopts with this? Maybe a redefining function?

```js
function _templateObject() {
  const result = _taggedTemplateLiteral(...);
  _templateObject = function() { return result; }
  return result;
}
```

We use the same pattern for [lazy-load imports](https://github.com/babel/babel/pull/6952/files#diff-ebca4dc345d93ae3ea13a65000a5d5acR3). /cc @loganfsmyth

Reply:
Can you elaborate a bit more on this? I don't understand what you mean by "function size deopts". :sob:

Reply:
> The redefining function for modules is needed because it is used in multiple places

I understand why we do it in imports. This is a different optimization using the same code pattern.

Old V8 (Crankshaft) has a character limit for inlining functions in JIT (even comments count towards this limit!):

```js
// I get called from a lot of places
function test() {
  // My function body is short, so I'll get inlined
  return tag(_templateObject, 1, 2, 3);
}

// But now, we lazy-load:
function test() {
  // We just inlined a really big template string expression,
  // so I cross the max_inlined_source_size limit
  // Guess this function won't be inlined anymore...
  return tag(_templateObject || (_templateObject = _taggedTemplateLiteral(['some really long string'])), 1, 2, 3);
}
```

Now, V8 switched to Turbofan in 5.9, so it's no big deal in recent releases. But, `Chrome < 59` and `Node < 8.3` have to worry about this. So, my idea is to reuse the same re-define function pattern to move the long template string out of the user's function, and get us back down to small function bodies.

Reply:
Thanks a lot for this explanation. 👍 I will work on implementing the redefining function over the weekend.